sdk/storage/azblob/internal/structuredmsg/structured_message.go

```diff
@@ -0,0 +1,369 @@
+			return nil, err
+		}
+	}
+
+	if r.segmentNum >= r.header.NumSegments {
+		return nil, io.EOF
+	}
+
+	// Read segment header
+	headerBytes := make([]byte, SegmentHeaderSize)
+	n, err := io.ReadFull(r.reader, headerBytes)
+	if err != nil {
+		if err == io.EOF {
+			return nil, ErrUnexpectedEOF
+		}
+		return nil, err
+	}
+	if n != SegmentHeaderSize {
+		return nil, ErrUnexpectedEOF
+	}
+
+	segmentNum := binary.LittleEndian.Uint32(headerBytes[0:])
+	dataLength := binary.LittleEndian.Uint32(headerBytes[4:])
+	expectedCRC64 := binary.LittleEndian.Uint64(headerBytes[8:])
+
+	if segmentNum != r.segmentNum {
+		return nil, ErrInvalidSegmentNumber
+	}
+
+	if dataLength > MaxSegmentSize {
+		return nil, ErrSegmentTooLarge
+	}
+
+	// Read segment data
+	data := make([]byte, dataLength)
+	n, err = io.ReadFull(r.reader, data)
+	if err != nil {
+		if err == io.EOF {
+			return nil, ErrUnexpectedEOF
+		}
+		return nil, err
+	}
+	if n != int(dataLength) {
+		return nil, ErrUnexpectedEOF
+	}
+
+	// Validate segment CRC64
+	actualCRC64 := crc64.Checksum(data, r.crc64Table)
+	if actualCRC64 != expectedCRC64 {
+		return nil, ErrCRC64Mismatch
+	}
+
+	// Update cumulative CRC64
+	r.totalCRC64 = crc64.Update(r.totalCRC64, r.crc64Table, data)
+	r.segmentNum++
+
+	return data, nil
+}
+
+// ReadAllSegments reads all remaining segments and returns them as a single byte slice
+func (r *StructuredMessageReader) ReadAllSegments() ([]byte, error) {
+	var allData []byte
+
+	for {
+		data, err := r.ReadSegment()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return nil, err
+		}
+		allData = append(allData, data...)
+	}
+
+	return allData, nil
+}
+
+// ReadTrailer reads and validates the message trailer
+func (r *StructuredMessageReader) ReadTrailer() (uint64, error) {
+	if r.closed {
+		return 0, io.EOF
+	}
+
+	// Validate segment count before reading trailer
+	if r.segmentNum != r.header.NumSegments {
+		return 0, ErrInvalidSegmentCount
+	}
+
+	trailerBytes := make([]byte, TrailerSize)
+	n, err := io.ReadFull(r.reader, trailerBytes)
+	if err != nil {
+		if err == io.EOF {
+			return 0, ErrUnexpectedEOF
+		}
+		return 0, err
+	}
+	if n != TrailerSize {
+		return 0, ErrUnexpectedEOF
+	}
+
+	// Validate trailer header
+	if trailerBytes[0] != MessageVersion {
+		return 0, ErrInvalidVersion
+	}
+	if trailerBytes[1] != FlagNone {
+		return 0, errors.New("invalid trailer flags")
+	}
+	properties := binary.LittleEndian.Uint16(trailerBytes[2:])
+	if properties&PropCRC64 == 0 {
+		return 0, errors.New("CRC64 property not set in trailer")
+	}
+
+	expectedCRC64 := binary.LittleEndian.Uint64(trailerBytes[4:])
+	if expectedCRC64 != r.totalCRC64 {
+		return 0, ErrCRC64Mismatch
+	}
+
+	r.closed = true
+	return expectedCRC64, nil
+}
+
+// GetTotalCRC64 returns the cumulative CRC64 of all segments read
+func (r *StructuredMessageReader) GetTotalCRC64() uint64 {
+	return r.totalCRC64
+}
+
+// Close marks the reader as closed
+func (r *StructuredMessageReader) Close() error {
```

Comment:
Is this really required?  We should only add an `io.Closer` when `Close()` will release underlying resources.